fem/pfespace.cpp

```diff
@@ -2390,6 +2390,206 @@ void ConformingProlongationOperator::MultTranspose(
+   Used in BuildNestedInterpolation(). */
+SparseMatrix * BuildLORP(
+   FiniteElementSpace& ho_fespace, FiniteElementSpace& lor_fespace,
+   const Array<int>& refine_relation)
+{
+   int ho_size = ho_fespace.GetVSize();
+   int lor_size = lor_fespace.GetVSize();
+   Mesh * lor_mesh = lor_fespace.GetMesh();
+
+   SparseMatrix * spmat = new SparseMatrix(lor_size, ho_size); // global P size
+   // loop over elements of fine, low-order mesh
+   for (int finee=0; finee<lor_mesh->GetNE(); ++finee)
+   {
+      int coarsee = refine_relation[finee];
+      Array<int> ho_dofs;
+      ho_fespace.GetElementDofs(coarsee, ho_dofs);
+      Array<int> lor_dofs;
+      lor_fespace.GetElementDofs(finee, lor_dofs);
+      DenseMatrix localP(lor_dofs.Size(), ho_dofs.Size());
+      for (int j=0; j<ho_dofs.Size(); ++j)
+      {
+         Vector v(ho_size); // global vector!
+         v = 0.0;
+         v(ho_dofs[j]) = 1.0;
+         GridFunction ho_vgf(&ho_fespace);
+         ho_vgf = v;
+         GridFunctionCoefficient ho_vgfc(&ho_vgf);
+         RefinedCoefficient ref_coeff(refine_relation, ho_vgfc);
+
+         Vector lor_proj(lor_dofs.Size());
+         lor_fespace.GetFE(finee)->Project(
+            ref_coeff, *lor_fespace.GetElementTransformation(finee), lor_proj);
+         for (int i=0; i<lor_dofs.Size(); ++i)
+         {
+            localP(i,j) = lor_proj[i];
+         }
+      }
+      spmat->SetSubMatrix(lor_dofs, ho_dofs, localP);
+   }
+   spmat->Finalize();
+   return spmat;
+}
+
+HypreParMatrix *BuildNestedInterpolation(ParFiniteElementSpace &ho_fespace,
+                                         ParFiniteElementSpace &lor_fespace)
+{
+   Array<int> refine_relation(lor_fespace.GetNE());
+   int multiplier = lor_fespace.GetNE() / ho_fespace.GetNE();
+   for (int i=0; i<ho_fespace.GetNE(); ++i)
+   {
+      for (int j=0; j<multiplier; ++j)
+      {
+         refine_relation[i*multiplier + j] = i;
+      }
+   }
+
+   SparseMatrix * mat;
+   if (ho_fespace.GetNVDofs() == 0) // Hdiv or Hcurl
+   {
+      mat = VectorBuildLORP(ho_fespace, lor_fespace, refine_relation);
+   }
+   else
+   {
+      mat = BuildLORP(ho_fespace, lor_fespace, refine_relation);
+   }
+
+   // construct the block-diagonal matrix A
+   HypreParMatrix *A = new HypreParMatrix(ho_fespace.GetComm(),
+                                          lor_fespace.GlobalVSize(),
+                                          ho_fespace.GlobalVSize(),
+                                          lor_fespace.GetDofOffsets(),
+                                          ho_fespace.GetDofOffsets(),
+                                          mat);
+
+   // this *adds* in some places where we want to *set*
+   HypreParMatrix *rap = RAP(lor_fespace.Dof_TrueDof_Matrix(), A,
+                             ho_fespace.Dof_TrueDof_Matrix());
+
+   // so we scale to account for that difference
+   Vector ones(lor_fespace.GetTrueVSize());
+   ones = 1.0;
+   lor_fespace.DivideByGroupSize(ones);
```

Comment:
That's not going to work with AMR (Groups are not available in NCMeshes).

Reply:
So far this is intended for meshes created with the LOR mesh constructor, which does not work if the HO mesh is non-conforming (segfault in pmesh.cpp line 890). 

What do you have in mind, to do AMR on both HO and LOR meshes after construction? Or to just do LOR on some portion of the domain, ie, adaptively? Some of these options are possible but will require significantly more work.

Reply:
You are absolutely right -- the LOR constructor currently also doesn't work for NC meshes, in fact making it work for that case will be non-trivial.

@jakubcerveny, it will be nice to think about adding this capability (LOR version of a HO NCMesh) in the future.